draft-ietf-opsawg-pcapng.md

```diff
@@ -1463,7 +1463,10 @@ following:
 | 0-1 | Inbound / Outbound packet (00 = information not available, 01 = inbound, 10 = outbound) |
 | 2-4 | Reception type (000 = not specified, 001 = unicast, 010 = multicast, 011 = broadcast, 100 = promiscuous). |
 | 5-8 | FCS length, in octets (0000 if this information is not available). This value overrides the if_fcslen option of the Interface Description Block, and is used with those link layers (e.g. PPP) where the length of the FCS can change during time. |
-| 9-15 | Reserved (MUST be set to zero). |
+| 9 | Csum NotReady, as a consequence the checksum field in the transport header is not valid, and checksum validation on the current packet should not be performed. |
+| 10 | Csum Valid, at least the transport header checksum of the packet has already been validated on the kernel side. If the flag is not set then we are free to check the checksum by ourselves provided that Csum NotReady is also not set. |
+| 11 | GSO TCP, for BIG TCP packets, the packet len field inside the IP header can be zero. In such case the actual network-level PDU length can be obtained from the Original Packet Length specified in the EPB. |
```

Comment:
That's not unique to Linux or to BIG TCP.  Code in Wireshark to deal with the total length being 0 in the IPv4 header dates back to at least 2012; it uses the packet length that was provided to libpcap and put into the packet record header as the total length in that case.

This flag should probably just be called the "TCP segmentation offloaded" flag, as there's no reason to presume it will only be used when capturing on an operating system that has something called, in the context of that OS, "GSO".

Reply:
Just curious why tcpdump doesn't use the packet length to deal with the total length being 0 in the IPv4 by default. Now it actually breaks when handling BIG TCP packets on tcpdump if GUESS_TSO is not enabled, while it works well in wireshark. is it because instead of trusting the packet length tcpudmp wants to handle it more carefully?

"TCP segmentation offloaded" looks better to me. Should I post v2 with your suggested words above? like:

| 9 | Checksum not ready, as a consequence of checksum offloading, e.g. a transmitted packet on an interface doing checksum offloading, so that the host networking stack doesn't compute and fill in the checksum before handing the packet either to the network adapter or the wraparound code path in the packet capture mechanism. |
| 10 | Checksum valid, the checksum has already been checked on the receive path before it was handed to the packet capture mechanism, so there's no need for the packet analyzer to check it. |
| 11 | TCP segmentation offloaded, this is either a received packet corresponding to several received link-layer packets, with reassembly having been done before the packet was handed to the packet capture mechanism, or a transmitted packet that will correspond to several link-layer packets after being fragmented, but that was wrapped around to the packet capture mechanism before the fragmentation occurred. |

Reply:
> Just curious why tcpdump doesn't use the packet length to deal with the total length being 0 in the IPv4 by default.

Because https://github.com/the-tcpdump-group/tcpdump/commit/3465ec4a0c94f3abe774dfdb9874f966403e20ce hasn't been backported to the 4.99 branch, so it's only in the main branch.  @fxlb, should that be backported to make captures with BIG TCP (and other forms of segmentation offloading) work better?  As noted, Wireshark already assumes segmentation offloading if the IP total length is 0.

Reply:
> should that be backported to make captures with BIG TCP (and other forms of segmentation offloading) work better?

Yes. I will backport the "IP: Enable TSO" update in the 4.99 branch (with some other recent fixes for the IP printer).

Reply:
> I will backport the "IP: Enable TSO" update in the 4.99 branch (with some other recent fixes for the IP printer).

Done.